### ISSUE TITLE
Trucks Stuck in Loop Removing and Delivering Sand to Ship Dock

### ISSUE THREAD

**Issue by utdrmac** [MaFi-Games/Captain-of-Industry-issues#253]
**Describe the issue**
Sand is being delivered to the dock (even when Keep Empty is checked). As soon as truck A empties sand into dock, after 1s, truck A fills sand back up, waits another 1-2s, then empties sand back into dock and drives away. Then empty truck B comes to dock, fills with sand, waits 1-2s, empties sand back to dock and drives away.

There is some strange, incorrect, demand-logic with sand and the dock. There is no request for sand at the dock, so why are the trucks bringing it in the first place? If 'Keep Empty' is checked, why are trucks bringing it to the dock?

Why does this only happen with sand? I'm not seeing this behavior with any other materials in the dock.

**Expected behavior**
* I expect the trucks to ignore delivering materials if the box 'keep empty' is checked.
* I expect some sort of filter on the dock that allows me to say 'this item can't be here', much like you can with mine control towers.

**To Reproduce**
View attached video of behavior (github required I zip it). Save game file also provided, although I saved after I took the video recording.

**Additional context**
 - Game version: 0.4.1e

**Attachments**
[TruckLovesSand.zip](https://github.com/MaFi-Games/Captain-of-Industry-issues/files/8857813/TruckLovesSand.zip)
[trucksSandStuck.zip](https://github.com/MaFi-Games/Captain-of-Industry-issues/files/8857817/trucksSandStuck.zip)

**Comment by Praefator** [comment]
Seems the issue has so far not re-occurred on 0.4.2a

**Comment by utdrmac** [comment]
I just woke up to that in the patch notes. Will take a look.

**Comment by utdrmac** [comment]
This is not fixed. The issue is far less severe than it was but I can still see trucks delivering sand to the docks, 2 units at a time, even though 'allow partial loads' is not checked.

**Comment by Honjura** [comment]
is there somewhere besides the dock to store the materials? if there is nowhere to store them then this will occur, this will also happen to an extent where u have raw materials such as ores in the dock but are also actively importing/mining these resources from other sources all trying to put the same resource into 1 destination, i.e supply of 3 demand of 1. 

After looking at the save few things.

1. Allow Partially loaded trucks is turned off.
2. the trucks are loading fully, then unloading fully, then loading partially.
 - the likely reasoning for this behaviour is because the Trucks are being told to load sand from the port, baseline behaviour is to keep the port empty where possible, so the trucks fully load sand, and realise their destination doesn't have the room for the sand, deposits back into the dock and then loads the correct amount to fill the storage without leftover.
 3. upgrading the storage for sand and setting the keep full to max empties the port
 4. after fixing this issue with the port by removing all sand, the trucks from the sand mine get about 4/5 of the way to the destination (Storage) and then realise the storage is full / doesn't have enough space for their load and turn back.
 5. assigned 3 more trucks to sand mine, filled them while sand storage was full, removed their assignment, and sand was dumped into port, this replicated the process of trucks constantly filling, emptying and partially filling from the port to try and fill the sand storage as it slowly emptied.
 
TLDR;
There is a limitation in the AI detecting storage capacities until they get close enough, the maps are large and the AI is doing its best. to help prevent this behaviour from occurring in the future, link storage from your mine control towers, this will keep trucks assigned to the mine tower updated on the storage levels and will prevent other sources from storing the material there while linked as a route.

**Comment by utdrmac** [comment]
> is there somewhere besides the dock to store the materials?

No. There is a bucket at the mine, and a bucket at cement processing. That's it. No other places for sand to go. Which only furthers the issue of "why are trucks taking sand to the dock" when there is literally only 1 place to take the sand (cement processing).

The dock has the highest P1 priority for emptying and the container at mine is P10. The AI should always be taking out of dock first. Once the dock is empty of sand, sand should never be taken there again.

> the trucks are loading fully, then unloading fully, then loading partially.

This was the pre 0.4.2 behavior. This loading-unloading loop seems to have been fixed, but a new root issue remains. Trucks are still taking sand to dock directly from the mine, but doing so in 1-2 unit increments despite partial-loads being disabled.

> upgrading the storage for sand and setting the keep full to max empties the port

This is a temporary solution. Once the bucket fills, sand will go from mine -> dock again.

> to help prevent this behaviour from occurring in the future, link storage from your mine control towers

I will verify this later today, but I'm pretty sure I already have this link in place and trucks still take minor amounts of sand to dock.

**Comment by JamieRCHI** [comment]
For me, I have noticed when deleting things that have stuff in or on them, like belts, the stuff has to go somewhere so the Dock is where it goes.

Would be helpful if the view of the Dock contents had a separate space with a label that shows that this is a temporary buffer for this. It should also have a higher priority than the other items that need to be placed in the dock when the ship returns and has stuff.

The dock needs separate storage for 
1. The items in the dock when you start the game.
2. Items returned to the dock when the ship returns and unloads.
3. A temporary buffer that places things there when things are deleted and there is no where else to place them.

Each section should have separate priority settings and options to move from one to the other.

**Comment by Praefator** [comment]
Honestly, having some way of dumping removed materials in _storages_ first (when storages of that item type exist) would save on a lot of micromanaging in manually placing several storages somewhere that will then need to be manually delivered to the places where they're actually useful. Shipyard in this case would be a secondary storage option for insta-removal **at best**

**Comment by zuff-coi** [comment]
This issue has been fixed. Trucks will no longer pick up goods that they have no storage to bring them to.

I loaded the save, the trucks that had the job to pick up sand did and then showed they had no place to deliver it. After building a storage for sand they dropped it off and stopped once the shipyard was full. 

![image](https://user-images.githubusercontent.com/108291414/179578268-75cabbdf-a12e-4154-a581-69e4f37ed402.png)

You can also straight up delete items in the Shipyard now, too, which would help with issues of overflowing cargo.

**Comment by zuff-coi** [comment]
This issue is still happening in v0.4.9c

**Comment by MaFi-Games** [comment]
I can no longer reproduce this issue and I believe we fixed it. Please file a new bug with a save file if this still occurs on latest version.
